src/PooledArrays.jl

```diff
@@ -121,6 +134,15 @@ large enough to hold all unique values in `array`.
 
 Note that if you hold mutable objects in `PooledArray` it is not allowed to modify them
 after they are stored in it.
+
+In order to improve performance of `getindex` and `copyto!` operations `PooledArray`s
+may share pools. This sharing is automatically undone by copying a shared pool before
+adding new values to it.
+
+It is not safe to assign values that are not already present in a `PooledArray`'s pool
+from one thread while either reading or writing to the same array from another thread
+(even if pools are not shared). However, reading and writing from different threads is safe
+if all values already exist in the pool.
```

Comment:
I wonder if eventually having a `FixedPooledArray` would help with this; a `FixedPooledArray` would have a "frozen" value pool and not allow new values on `setindex!`. Then for critical operations, you could "freeze" a PooledArray's pool when doing operations or something.

Reply:
Would it really help compared with what we have now? You can already freeze the pool: you just have to avoid adding any new values to it. :-) Freezing would just allow catching bugs, but we could implement this for `PooledArray` if we wanted to.

Reply:
@nalimilan - since we disallow adding levels in threaded code anyway do we really need copy-on-write behavior? It would be simpler to never copy (the original reason for introducing copy-on-write was safeguarding against problems with threading).

Then https://github.com/JuliaData/PooledArrays.jl/issues/58 would provide a way to force performing a copy if needed (most likely in cases when the original pool was very large and it is no longer needed).

Reply:
We disallow adding levels to a `PooledArray` in one thread while doing other operations on the same array in another, but it wouldn't be reasonable to disallow adding levels to any `PooledArray` in a threaded context IMO. That's why I think it's good to undo pool sharing before adding levels.

Reply:
OK - we can leave it. Keeping track of `refcount` is not expensive.